pkg/autoscaler/config.go

```diff
@@ -158,10 +150,36 @@ func NewConfigFromMap(data map[string]string) (*Config, error) {
 		}
 	}
 
+	return validate(lc)
+}
+
+func validate(lc *Config) (*Config, error) {
 	if lc.ScaleToZeroGracePeriod < 30*time.Second {
 		return nil, fmt.Errorf("scale-to-zero-grace-period must be at least 30s, got %v", lc.ScaleToZeroGracePeriod)
 	}
 
+	if lc.ContainerConcurrencyTargetFraction <= 0 || lc.ContainerConcurrencyTargetFraction > 1 {
+		return nil, fmt.Errorf("container-concurrency-target-percentage = %f is outside of valid range of (0, 100]", lc.ContainerConcurrencyTargetFraction)
+	}
+
+	if x := lc.ContainerConcurrencyTargetFraction * lc.ContainerConcurrencyTargetDefault; x < 1.0 {
+		return nil, fmt.Errorf("container-concurrency-target-percentage and container-concurrency-target-default yield target concurrency of %f, can't be less than 1", x)
+	}
+
+	// We can't permit stable window be less than our aggregation window for correctness.
+	if lc.StableWindow < BucketSize {
```

Comment:
I don't understand why we can't accept something smaller - it's not a guarantee, it's a desire/best-effort. So in this impl our best-effort will result in "BucketSize".  But it would be nice if the config parameters could be portable across multiple implementations of Kn (as best we can anyway).

Reply:
Additionally, forcing people to know that the minimum is 2 seconds is exposing too much of the underlying implementation. People need a way to say "terminate the instance as soon as possible when it's idle" w/o having to work too hard to figure out what the minimum value can be.  Just let "0" be that value.

Reply:
Because the system algorithm as we have now, does not work with smaller number.
It is unreasonable to expect system to act faster than its control loop time.  It does not behave as "best-effort" it rather breaks in unexpected ways.
You have discovered that first hand. :-)

Reply:
„stable-window“ is not the value that determines when your pods will be idle. It‘s only implicitly so because that‘s how the math behind the aggregation works. „stable-window“ in itself already exposes the aggregation-style behavior of the autoscaler.

On top of that: This is an annotation of the KPA specifically. It‘s not part of the spec and thus not part of conformance testing. Nothing guarantees other implementations even respect it.

I would agree if this knob was „pod-idle-shutdown-period“, but that‘s not the case neither is this knob geared towards being that.

Reply:
> Additionally, forcing people to know that the minimum is 2 seconds is exposing too much of the underlying implementation. People need a way to say "terminate the instance as soon as possible when it's idle" w/o having to work too hard to figure out what the minimum value can be. Just let "0" be that value.

Sure, please file an issue, since it would require additional implementation. This safeguards the system from operating outside of its designed parameters.

Reply:
> Because the system algorithm as we have now, does not work with smaller number.
> It is unreasonable to expect system to act faster than its control loop time. It does not behave as "best-effort" it rather breaks in unexpected ways.
> You have discovered that first hand. :-)

I don't think I'm being clear. We should accept a lower value but then in the code treat it like 2. This way we don't require the person managing the system to have to understand too much or figure out what value is the minimum via errors.

> Sure, please file an issue, since it would require additional implementation. This safeguards the system from operating outside of its designed parameters.

My original issue is that issue  :-)

> I would agree if this knob was „pod-idle-shutdown-period“, but that‘s not the case neither is this knob geared towards being that.

Perhaps we need one then. From a UX perspective there is a lot of value in keeping it simple for admins, not just devs.  As an example, yesterday you told me that I had to add two values to determine how long before an instance will be terminated... I think asking most people to understand that level of detail is asking too much. We should offer a higher-level config option (per the spec) and then each impl can map that to impl config options that are (mostly) hidden, until someone choose to dive deep.

Reply:
By accepting a lower value and silently treating as a higher we're introducing a non-intuitive user experience, where their actions don't translate into system behavior. I'd personally be quite annoyed...

Reply:
But we do this already with just about all other values w.r.t. scaling - they're not guarantees, they're lower-bounds. E.g. the 90 idle timeout is not something I see that often, usually it often more than 90 seconds.

Reply:
> As an example, yesterday you told me that I had to add two values to determine how long before an instance will be terminated...
You don't unless you want to _configure_ those values. Defaults are meant to be sensible. If they aren't we should probably tune them.

I do not disagree that 0 config, or simpler config is better than what we have now. But first we need to make the rest of the system more self-driving and requiring less tuning (to that end I am drafting some notes), but we're not there yet, and for now we need to make sure that values provided to the system, will translate to corresponding behavior (especially that some values are derivative of other numbers via % application). All we do is steps towards that bright future — but it's not a single step, unfortunately.

If you have good improvements and simplifications that reduce the number of configuration values we need, I'd be more than happy to review the PRs.